## Session #1 Homework

### Set up the environment

You need to install Python, NumPy, Pandas, Matplotlib and Seaborn. For that, you can the instructions from
[06-environment.md](https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/01-intro/06-environment.md).

### Question 1

What's the version of NumPy that you installed? 

You can get the version information using the `__version__` field:

```python
np.__version__
```

In [4]:
import numpy as np
import pandas as pd

np.__version__

'1.22.3'


### Getting the data 

For this homework, we'll use the Car price dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv).

You can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
```

Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [7]:
# df = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv")
df = pd.read_csv("data/car_price.csv")
df

Unnamed: 0     Make       Model  Year                Engine Fuel Type  \
0               0      BMW  1 Series M  2011     premium unleaded (required)   
1               1      BMW    1 Series  2011     premium unleaded (required)   
2               2      BMW    1 Series  2011     premium unleaded (required)   
3               3      BMW    1 Series  2011     premium unleaded (required)   
4               4      BMW    1 Series  2011     premium unleaded (required)   
...           ...      ...         ...   ...                             ...   
11909       11909    Acura         ZDX  2012     premium unleaded (required)   
11910       11910    Acura         ZDX  2012     premium unleaded (required)   
11911       11911    Acura         ZDX  2012     premium unleaded (required)   
11912       11912    Acura         ZDX  2013  premium unleaded (recommended)   
11913       11913  Lincoln      Zephyr  2006                regular unleaded   

       Engine HP  Engine Cylinders Transmission Type      Driven_Wheels  \
0          335.0               6.0            MANUAL   rear wheel drive   
1          300.0               6.0            MANUAL   rear wheel drive   
2          300.0               6.0            MANUAL   rear wheel drive   
3          230.0               6.0            MANUAL   rear wheel drive   
4          230.0               6.0            MANUAL   rear wheel drive   
...          ...               ...               ...                ...   
11909      300.0               6.0         AUTOMATIC    all wheel drive   
11910      300.0               6.0         AUTOMATIC    all wheel drive   
11911      300.0               6.0         AUTOMATIC    all wheel drive   
11912      300.0               6.0         AUTOMATIC    all wheel drive   
11913      221.0               6.0         AUTOMATIC  front wheel drive   

       Number of Doors                        Market Category Vehicle Size  \
0                  2.0  Factory Tuner,Luxury,High-Performance      Compact   
1                  2.0                     Luxury,Performance      Compact   
2                  2.0                Luxury,High-Performance      Compact   
3                  2.0                     Luxury,Performance      Compact   
4                  2.0                                 Luxury      Compact   
...                ...                                    ...          ...   
11909              4.0             Crossover,Hatchback,Luxury      Midsize   
11910              4.0             Crossover,Hatchback,Luxury      Midsize   
11911              4.0             Crossover,Hatchback,Luxury      Midsize   
11912              4.0             Crossover,Hatchback,Luxury      Midsize   
11913              4.0                                 Luxury      Midsize   

       Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
0              Coupe           26        19        3916  46135  
1        Convertible           28        19        3916  40650  
2              Coupe           28        20        3916  36350  
3              Coupe           28        18        3916  29450  
4        Convertible           28        18        3916  34500  
...              ...          ...       ...         ...    ...  
11909  4dr Hatchback           23        16         204  46120  
11910  4dr Hatchback           23        16         204  56670  
11911  4dr Hatchback           23        16         204  50620  
11912  4dr Hatchback           23        16         204  50920  
11913          Sedan           26        17          61  28995  

[11914 rows x 17 columns]

### Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

- 16
- 6572
- 11914
- 18990

In [8]:
df.shape

(11914, 17)

### Question 3

Who are the most frequent car manufacturers (top-3) according to the dataset?

- Chevrolet, Volkswagen, Toyota
- Chevrolet, Ford, Toyota
- Ford, Volkswagen, Toyota
- Chevrolet, Ford, Volkswagen

> **Note**: You should rely on "Make" column in this question.

In [7]:
df.Make.value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

### Question 4

What's the number of unique Audi car models in the dataset?

- 3
- 16
- 26
- 34

In [9]:
df.query("Make=='Audi'")["Model"].nunique()

34

### Question 5

How many columns in the dataset have missing values?

- 5
- 6
- 7
- 8

In [21]:
np.array([df.isna().sum() > 0]).sum()

5

### Question 6

1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the `fillna` method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

> Hint: refer to existing `mode` and `median` functions to complete the task.

- Yes
- No

In [23]:
df["Engine Cylinders"].median()

6.0

In [30]:
df["Engine Cylinders"].mode()

0    4.0
Name: Engine Cylinders, dtype: float64

In [33]:
df["Engine Cylinders"].fillna(df["Engine Cylinders"].mode()).median()

6.0

### Question 7

1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using `drop_duplicates` method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the value of the first element of `w`?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- -0.0723
- 4.5949
- 31.6537
- 63.5643

In [40]:
X = df.query("Make == 'Lotus'")[["Engine HP", "Engine Cylinders"]].drop_duplicates().to_numpy()
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [49]:
XTX = X.T.dot(X)

In [50]:
np.linalg.inv(XTX)

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [51]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [52]:
w = np.linalg.inv(XTX).dot(X.T).dot(y)
w 

array([  4.59494481, -63.56432501])

## Submit the results

Submit your results here: https://forms.gle/vLp3mvtnrjJxCZx66

If your answer doesn't match options exactly, select the closest one.